test/e2e/set/scyllacluster/scyllacluster_shardawareness.go

```diff
@@ -0,0 +1,114 @@
+
+	g.It("should allow to build connection pool using shard aware ports", func() {
+		const (
+			nonShardAwarePort = 9042
+			shardAwarePort    = 19042
+			nrShards          = 2
+		)
+
+		ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+		defer cancel()
+
+		sc := scyllaclusterfixture.BasicScyllaCluster.ReadOrFail()
+		sc.Spec.Datacenter.Racks[0].Members = 1
+
+		// Ensure 2 shards.
+		sc.Spec.Datacenter.Racks[0].Resources.Limits[corev1.ResourceCPU] = resource.MustParse(fmt.Sprintf("%d", nrShards))
+
+		framework.By("Creating a ScyllaCluster")
+		err := framework.SetupScyllaClusterSA(ctx, f.KubeClient().CoreV1(), f.KubeClient().RbacV1(), f.Namespace(), sc.Name)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		sc, err = f.ScyllaClient().ScyllaV1().ScyllaClusters(f.Namespace()).Create(ctx, sc, metav1.CreateOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		framework.By("Waiting for the ScyllaCluster to deploy")
+		waitCtx1, waitCtx1Cancel := utils.ContextForRollout(ctx, sc)
+		defer waitCtx1Cancel()
+		sc, err = utils.WaitForScyllaClusterState(waitCtx1, f.ScyllaClient().ScyllaV1(), sc.Namespace, sc.Name, utils.IsScyllaClusterRolledOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		hosts, err := utils.GetHosts(ctx, f.KubeClient().CoreV1(), sc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		connections := map[uint16]string{}
+		connectionsMut := sync.Mutex{}
+
+		clusterConfig := gocql.NewCluster(hosts...)
+		clusterConfig.Dialer = DialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
+			sourcePort := gocql.ScyllaGetSourcePort(ctx)
+			localAddr, err := net.ResolveTCPAddr(network, fmt.Sprintf(":%d", sourcePort))
+			if err != nil {
+				return nil, err
+			}
+
+			framework.Infof("Connecting to %s using %d source port", addr, sourcePort)
+			connectionsMut.Lock()
+			connections[sourcePort] = addr
+			connectionsMut.Unlock()
+
+			d := &net.Dialer{LocalAddr: localAddr}
+			return d.DialContext(ctx, network, addr)
+		})
+
+		framework.By("Waiting for the driver to establish connection to shards")
+		session, err := gocqlx.WrapSession(clusterConfig.CreateSession())
+		o.Expect(err).NotTo(o.HaveOccurred())
+		defer session.Close()
+
+		err = wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+			return len(connections) == nrShards, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		shardAwareAttempts := 0
+		for sourcePort, addr := range connections {
+			// Control connection is also put in pool, and it always uses the default port.
+			if sourcePort == 0 {
+				o.Expect(addr).To(o.HaveSuffix(fmt.Sprintf("%d", nonShardAwarePort)))
+				continue
+			}
+			o.Expect(addr).To(o.HaveSuffix(fmt.Sprintf("%d", shardAwarePort)))
+			shardAwareAttempts++
+		}
+
+		// Control connection always use the default port to discover the shard number,
+		// and is added to the pool with shard number it randomly gets.
+		o.Expect(shardAwareAttempts).To(o.Equal(nrShards - 1))
```

Comment:
should there be 2 shard aware attempts? this assert just 1

Reply:
No, because first connection is always to non-shard aware port to discover the number of shards. There's a comment about it above

Reply:
What is not clear to me is that if you use the first connection to discover the number of shards, why doesn't it connect to each of them afterwards?

Reply:
Because it's already established to random shard. There's no point to do termination and roundtip again if only missing ones can be established.